### ISSUE TITLE
Implement `str` namespace functions on categoricals

### ISSUE THREAD

**Issue by mcrumiller** [pola-rs/polars#9773]
### Problem description

All [string functions in the `str` namespace](https://pola-rs.github.io/polars/py-polars/html/reference/series/string.html) should be able to work extremely fast on categorical columns by manipulating the labels alone and expanding the result. However, right now categoricals must first be cast to `pl.Utf8` and then cast back, which is wasteful:

```python
import polars as pl

s = pl.Series(["test", "hello", "hello", "hello", "test"], dtype=pl.Categorical)

# current requirement
s2 = s.cast(pl.Utf8).str.slice(1).cast(pl.Categorical)

# proposed
s2 = s.str.slice(1)
```

**Comment by deanm0000** [comment]
I wonder if this is more efficient than round tripping the cast on all the values...

```
s.to_frame().join(
    s.to_frame().unique().with_columns(pl.col('').cast(pl.Utf8()).str.slice(1).cast(pl.Categorical()).alias('new')), on='').get_column('new').alias('')
```

I imagine that's roughly how to implement the feature, although I'd guess you probably want to do it inside a StringCache but not really sure.

**Comment by stevenlis** [comment]
Hi @mcrumiller Is this still planned for categoricals and enums?

**Comment by ritchie46** [comment]
I am really not sure about this. I don't think categorical should behave like strings semantically. We also first need refactor our categoricals to be agnostic of location, before we could even do all this.

But again, I don't think we should.

**Comment by ritchie46** [comment]
> it's a shame to not be able to leverage that efficiency for string operations

Then we should improve efficiency on the string storage. We can intern the strings with views. We should not misuse a data type that is not meant for that and needs global synchronization of categories. This global synchronizations cause already enough complication as it is. Applying expressions that change your datatype depending on the data, should be minimized at all cost.

**Comment by mcrumiller** [comment]
Ok, I have come around and agree with you about string ops that change the categories.

@ritchie46 what about queries about the values? Going through the `str` API, I see the following that I think would be reasonable to implement on categoricals using a fast path:

* `contains`/`contains_any`
* `count_matches`
* `decode`/`encode`
* `starts_with`/`ends_width`
* `find`/`find_many`
* `len_bytes`/`len_chars`
* `to_date`/`to_datetime`/`to_<x>`

If you think this is reasonable, I can open a new issue and use some of my existing work for these.

**Comment by ritchie46** [comment]
We could discuss if those methods fit in the categorical namespace. Which I think some do. I don't want to implement them independently though. We should be able to call these methods on the categories only and  then use the categories to materialize. This can all use the same code path as it is essentially a `str.method` + `gather`, and thus would not require any code bloat.

**Comment by mcrumiller** [comment]
Yep--that's what I had done for `head`/`tail`/`slice`--they went through `str.<method>` on the categories alone, but required a remap too, which I can tell we don't want. I'll put something together for `len_bytes`/`len_chars` first (something simple). I think `contains`/`starts_with`/`ends_with` and `len_bytes`/`len_chars` are probably the only methods above that people would care about, possibly in addition to the `to_<x>` methods which have an obvious fast path.

I'll close this and perhaps open a new issue for a much smaller subset of methods that would make sense to have a simple fast path. Thanks for the feedback @ritchie46.

**Comment by mcrumiller** [comment]
Reopening with a better solution.
